src/main/scala/com/fulcrumgenomics/bam/FindSwitchbackReads.scala

```diff
@@ -0,0 +1,395 @@
+    val iterator = Bams.templateIterator(in)
+
+    val outHeader = in.header.clone()
+    sortOrder match {
+      case Some(so) => so.applyTo(outHeader)
+      case None     =>
+        if (in.header.getSortOrder != SortOrder.queryname && in.header.getGroupOrder != GroupOrder.query) {
+          SamOrder.Queryname.applyTo(outHeader)
+        }
+    }
+    sortOrder.foreach(_.applyTo(outHeader))
+    val out = SamWriter(output, header=outHeader, sort=sortOrder)
+
+    logger.info("Examining reads for switchbacks.")
+    iterator.map { template =>
+      total += 1
+      if (!template.primaryReads.exists(_.mapped)) template else {
+        aligned += 1
+
+        val result: Option[SwitchHit] = findReadBasedSwitchback(refs, template, minLength, maxOffset, maxErrorRate)
+          .orElse(findTandemSwitchback(template, maxGap))
+
+        result match {
+          case None =>
+            Unit
+          case Some(hit: TandemBasedHit) =>
+            switchbackGaps.count(hit.gap)
+            val tagValue = s"${SwitchTypeTandem},${hit.gap}"
+            template.allReads.foreach(r => r(SwitchTag) = tagValue)
+          case Some(hit: ReadBasedHit) =>
+            switchbackLengths.count(hit.length)
+            switchbackOffsets.count(hit.offset)
+            val foundInReadNum = readNum(hit.read.getOrElse(unreachable("read should be populated!")))
+            template.allReads.foreach { r =>
+              val num = readNum(r)
+              r(SwitchTag) = s"${SwitchTypeRead},${if (num == foundInReadNum) "read" else "mate"},${hit.offset},${hit.length}"
+            }
+        }
+
+        if (result.isDefined && !this.dontUnmap) template.unmapped else template
+      }
+    }
+    .foreach { template =>
+      template.allReads.foreach { rec =>
+        out += rec
+        progress.record(rec)
+      }
+    }
+
+    in.safelyClose()
+    out.close()
+
+    // Output the metrics
+    this.metrics.foreach { prefix =>
+      val m = SwitchMetric(
+        sample                   = summarizeRg(header=in.header, f = _.getSample),
+        library                  = summarizeRg(header=in.header, f = _.getLibrary),
+        templates                = total,
+        aligned_templates        = aligned,
+        switchback_templates     = switchbackLengths.total + switchbackGaps.total,
+        fraction_switchbacks     = (switchbackLengths.total + switchbackGaps.total) / aligned.toDouble,
+        read_based_switchbacks   = switchbackLengths.total,
+        mean_length              = switchbackLengths.mean(),
+        mean_offset              = switchbackOffsets.mean(),
+        tandem_based_switchbacks = switchbackGaps.total,
+        mean_gap                 = switchbackGaps.mean()
+      )
+
+      Metric.write(PathUtil.pathTo(prefix + ".summary.txt"), m)
+
+      for ((suffix, counter, header) <- Seq(("lengths", switchbackLengths, "length"), ("offsets", switchbackOffsets, "offset"), ("gaps", switchbackGaps, "gap_length"))) {
```

Comment:
just a style thing, but 
```suggestion
Seq(
    ("lengths", switchbackLengths, "length"), 
    ("offsets", switchbackOffsets, "offset"),
    ("gaps", switchbackGaps, "gap_length")
).foreach { case suffix, counter, header) =>
```

Reply:
Just a style thing for me, since now it is much easier to see how things align